### PR TITLE
Leverage AST deltas for narrowing action scopes.

### PR DESCRIPTION
**Pull request by isidentical**
This PR implements a token-level position provider for definitions (functions, classes) and an action optimizer that can transform `Replace`'s into `_Rename`s in order to only change the name of a function if there are no other changes.

### CODE REVIEW

File: refactor/internal/ast_delta.py [hunk 1]
```diff
@@ -0,0 +1,150 @@
+import ast
+from contextlib import suppress
+from dataclasses import dataclass
+from enum import Enum, auto
+from functools import cache, partial
+from typing import Any, Dict, Iterator, List, Optional, Type, cast
+
```
Comment (hakancelikdev) [inline refactor/internal/ast_delta.py:7]:
```
__all__
```

File: refactor/internal/ast_delta.py [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+            on_field=field,
+            on_index=index,
+        )
+        with suppress(_Continue):
+            if isinstance(base_item, ast.AST) or base_item is None:
+                if new_item is None or base_item is None:
+                    yield from _item_change_if(
+                        new_item is not base_item, ChangeType.FULL
+                    )
+                    continue
```
Comment (hakancelikdev) [inline refactor/internal/ast_delta.py:143]: Using continue statement seems unnecessary.

File: refactor/internal/position_provider.py [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+import tokenize
+from functools import singledispatch
+from itertools import chain
+from typing import Callable, Iterator, Optional, Union
+
+from refactor import common
+from refactor.ast import Lines, split_lines
+from refactor.context import Context
+
+
```
Comment (hakancelikdev) [inline refactor/internal/position_provider.py:11]:
```
__all__
```

File: refactor/internal/ast_delta.py (hunk 2 above)
Comment (isidentical) [reply refactor/internal/ast_delta.py]: Good catch!

**Pull request merged**